bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java

```diff
@@ -0,0 +1,82 @@
+        }
+
+        Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap = typeParameterMap.getPropertyToClassParamIndexMap();
+        Either<Integer, TypeParameterMap> classTypeParamRepresentsWholeField = propertyToClassParamIndexMap.get(-1);
+        if (classTypeParamRepresentsWholeField != null) {
+            Integer index = classTypeParamRepresentsWholeField.map(i -> i, e -> {
+                throw new IllegalStateException("Invalid state, the whole class cannot be represented by a subtype.");
+            });
+            return (TypeData<V>) typeParameters.get(index);
+        } else {
+            return getTypeData(typeData, typeParameters, propertyToClassParamIndexMap);
+        }
+    }
+
+    private static <V> TypeData<V> getTypeData(final TypeData<V> parentTypeData, final List<TypeData<?>> specializedTypeParameters,
+                                               final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap) {
+        List<TypeData<?>> typeParameters = new ArrayList<>(parentTypeData.getTypeParameters());
+        List<TypeData<?>> subTypeParameters = new ArrayList<>();
+
+        for (int i = 0; i < typeParameters.size(); i++) {
+            subTypeParameters.add(getTypeData(typeParameters.get(i), specializedTypeParameters, propertyToClassParamIndexMap, i));
+        }
+        return TypeData.builder(parentTypeData.getType()).addTypeParameters(subTypeParameters).build();
+    }
+
+    private static TypeData<?> getTypeData(final TypeData<?> parentTypeData, final List<TypeData<?>> specializedTypeParameters,
+                                           final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap,
+                                           final int index) {
+        if (!propertyToClassParamIndexMap.containsKey(index)) {
+            return parentTypeData;
+        }
+        return propertyToClassParamIndexMap.get(index).map(l -> {
+                    if (parentTypeData.getTypeParameters().isEmpty()) {
+                        // Represents the whole typeData
+                        return specializedTypeParameters.get(l);
+                    } else {
+                        // Represents a single nested type parameter within this typeData
+                        TypeData.Builder<?> builder = TypeData.builder(parentTypeData.getType());
+                        List<TypeData<?>> typeParameters = new ArrayList<>(parentTypeData.getTypeParameters());
+                        typeParameters.set(index, specializedTypeParameters.get(l));
+                        builder.addTypeParameters(typeParameters);
+                        return builder.build();
+                    }
+                },
+                r -> {
+                    // Represents a child type parameter of this typeData
+                    return getTypeData(parentTypeData, specializedTypeParameters, r.getPropertyToClassParamIndexMap());
```

Comment:
The `parentTypeData` doesn't need specializing but one of its nested type parameters does, so we recurse down a level.